Take ownership of jdk.internal.misc.Unsafe

theresa-m opened the pull request:
Take ownership of jdk.internal.misc.Unsafe library to support
narrow primitive types (byte, short, boolean, char).

Signed-off-by: Theresa Mammarella <Theresa.T.Mammmarella@ibm.com>

Test PR: #524 

Reviewer: @babsingh

theresa-m commented:
Testing in progress.

genie-openj9 commented:
Can one of the admins verify this patch?

genie-openj9 commented:
Can one of the admins verify this patch?

DanHeidinga commented:
We should take https://github.com/eclipse/openj9/pull/210 into account when writing `ensureClassIsInitialized()`.

Also, we'll want a JIT team review on this once we've got complete implementation as there will be opportunities to get better perf by using JITHelpers to implement parts of this.

FYI @andrewcraik

theresa-m commented:
Supporting tests: https://github.com/eclipse/openj9/pull/524

babsingh commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,2653 @@
+
+	public final long getLongUnaligned(Object obj, long offset, boolean bigEndian) {
+		long result = getLongUnaligned(obj, offset);
+		return convEndian(bigEnd
```
initialize `short result = 0;`

babsingh replied:
`((obj == null) || obj.getClass().isArray()) {`

babsingh replied:
comments on newline:
```
} else { 
    /* object is non-null primitive type */
```

theresa-m commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,2653 @@
+
+	public final int getAndAddInt(Object obj, long index, int value) {
+		for (;;) {
+			int intAtIndex = getIntVolatile(obj, index);
```
`weakCompareAndSetInt` should only fail if the passed in compareValue is not the same as what is written in memory which could occur since `getAndAddInt` is not atomic. If it fails we would need to try again with the new value stored in memory so updating inside the loop is necessary.

theresa-m commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,2653 @@
+		return theUnsafe;
+	}
+
+	public native int getInt(Object obj, long offset);
```
Confirmed all methods should have these with Dan. I'll take another pass and add these in.

pdbain-ibm commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,2724 @@
+			boolean result = weakCompareAndSetLongRelease(obj, index, longAtIndex, longAtIndex & value);
+			if (result) {
+				return longAtIndex;
+			
```
Exceptions with messages should use com.ibm.oti.util.Msg.getString() for national language support (NLS).

pdbain-ibm commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,2724 @@
+
+	public Object allocateUninitializedArray(Class<?> c, int length) {
+		if (null == c) {
+			throw new IllegalArgumentException("Component type is null");
```
Use com.ibm.oti.util.Msg.getString() here too.

pdbain-ibm commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,2724 @@
+		}
+
+		if (length < 0) {
+			throw new IllegalArgumentException("Negative length");
```
And here.

theresa-m commented:
@babsingh I added javadocs to the class. Can you take a look when you have a chance? I split the commits up a bit to hopefully make reviewing a bit easier.

pdbain-ibm commented:
Is this ready for merging?  https://github.com/eclipse/openj9/pull/579 requires this change.

babsingh commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6206 @@
+     * @param offset specifies where to compare and store value in obj
+     * @param compareValue value that is expected to be in obj at offset
+     * @param setValue value that will be set in obj at offset if compare is successful
+     * @return boolean value indicatin
```
extra line

babsingh replied:
extra line

babsingh replied:
extra line

babsingh replied:
at multiple places, the spacing is inconsistent

babsingh replied:
extra line

babsingh replied:
extra line

babsingh replied:
extra line

babsingh replied:
extra line

babsingh replied:
extra line

pdbain-ibm replied:
Extra spaces.

pdbain-ibm replied:
It's marginally more efficient if you inline the call into the if statement (avoids storing and loading a local).

pdbain-ibm replied:
return value != 0; 
would be clearer.

babsingh commented on runtime/jcl/common/sun_misc_Unsafe.cpp:
```diff
@@ -677,11 +677,26 @@ Java_sun_misc_Unsafe_getKlassPointer(JNIEnv *env, jobject receiver, jobject addr
-	return JNI_FALSE;
+	jboolean result = JNI_FALSE;
+
+	J9VMThread *currentThread = (J9VMThread*)env;
```
`(J9VMThread *)env`

pdbain-ibm commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6206 @@
+	
+	/* For long shift instructions (lshr, lushr, lshl) only the low 6 bits of the
+	 * shift amount s are considered
+	 */
```
Ibid.

pdbain-ibm commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6206 @@
+	 * Gets the value of the integer in the obj parameter referenced by offset.
+	 * This is a non-volatile operation.
+	 * 
+	 * @param obj object in which to retrieve the value
```
I would say "from which" throughout.

pdbain-ibm commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6206 @@
+	 * This is a non-volatile operation.
+	 * 
+	 * @param obj object in which to retrieve the value
+	 * @param offset specifies where to retrieve value in obj
```
Better phrased as "position of the value in obj".

pdbain-ibm commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6206 @@
+	 * This is a non-volatile operation.
+	 * 
+	 * @param obj object in which to store the value
+	 * @param offset specifies where to store value in obj
```
See above.

pdbain-ibm commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6206 @@
+	 * This is a non-volatile operation.
+	 * 
+	 * @param obj object in which to retrieve the value
+	 * @param offset specifies where to retrieve value in obj
```
Same comment throughout: remove "specifies".

pdbain-ibm commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6206 @@
+	 * requested size plus space for two pointers.
+	 *
+	 * @param size number of bytes to allocate
+	 * @return address to allocated buffer
```
address of allocated... would be better.

pdbain-ibm commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6206 @@
+	 * Reallocates a block and adds it to a doubly-linked list for tracking purposes.  Actual size of the block is
+	 * requested size plus space for two pointers.
+	 * 
+	 * @param address to old buffer
```
ibid.

pdbain-ibm commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6206 @@
+	 * 			valid class file as defined by The JVM Spec
+	 * @param offset offset of the start of the class data in b
+	 * @param bLength length of the class data
+	 * @param cl used to load the class being built. If null, the default system ClassLoader will be used
```
Did you mean "@param cl **classloader** used ..."?

pdbain-ibm commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6206 @@
+     * The get operation has memory semantics of getVolatile.
+     * The set operation has the memory semantics of setVolatile.
+     *
+     * @param obj object in which to store the value
```
I suggest "into which" (here and elsewhere).

pdbain-ibm commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6206 @@
+     * @param offset specifies where to compare and store value in obj
+     * @param compareValue value that is expected to be in obj at offset
+     * @param setValue value that will be set in obj at offset if compare is successful
+     * @return boolean value indicatin
```
Please fix the javadoc format, i.e. "/**<newline>".

theresa-m replied:
Thanks for all the feedback! I've left private methods without a javadoc and just a normal comment since they won't be accessed outside the class.

pdbain-ibm commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6206 @@
+	 * @throws NullPointerException if class is null
+	 */
+	public boolean shouldBeInitialized(Class<?> c) {
+		if (null == c) {
```
I suggest using `Objects.requireNotNull(c),` here and throughout.

theresa-m commented:
@DanHeidinga or @pshipton this is ready for a senior review

pshipton commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6138 @@
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code is also Distributed under one or more Secondary Licenses,
```
This isn't the correct license, please update it from another OpenJ9 file.

pshipton commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6138 @@
+	public static final int INVALID_FIELD_OFFSET;
+	
+	/**
+	 * Starting offset of Boolean array
```
I don't think any of the native type names should be capitalized, as they could be confused with objects like java.lang.Boolean.

pshipton commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6138 @@
+	/* number of bits in a byte */
+	private static final int BITS_IN_BYTE = 8;
+
+	/* 32 bit mask */
```
to be consistent, the comment should include "(long")

pshipton commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6138 @@
+		UNALIGNED_ACCESS = theUnsafe.unalignedAccess0();
+	}
+
+	/* [IF ] */
```
Is there a reason to remove the comment from the preprocessed source?

theresa-m replied:
`registerNatives()` was defined this way in some other jcl classes but its not necessary. I've updated the comment.

pshipton commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6138 @@
+	static native void registerNatives();
+	
+	/**
+	 * Gets the value of the integer in the obj parameter referenced by offset.
```
Call it what it is, "int" instead of "integer"

pshipton commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6138 @@
+	public native Object allocateInstance(Class<?> c) throws InstantiationException;
+
+	/**
+	 * Execute throwable parameter.
```
Should this be "Throw" rather than "Execute"?

pshipton commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6138 @@
+	public native void putObjectVolatile(Object obj, long offset, Object value);
+
+	/**
+	 * Atomically gets the value of the integer in the obj parameter referenced by offset.
```
int instead of integer

theresa-m commented:
Thanks @pshipton. I've included your comments into my commit.

pshipton commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6134 @@
+	 * Get the load average in the system.
+	 * 
+	 * @param loadavg array of elements
+	 * @param numberOfElements number of samples
```
numberOfElements doesn't match the parameter name nelems

pshipton commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6134 @@
+	 * 
+	 * @param obj object into which to store the address
+	 * position of the value in objstore value in obj
+	 * @param value address to store in obj
```
The address parameter is missing

pshipton replied:
Same for putByte(), putShort(), etc.

pshipton commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6134 @@
+	/**
+	 * Allocates a block and adds it to a doubly-linked list for tracking purposes.  Actual size of the block is
+	 * requested size plus space for two pointers.
+	 * If size passed is 0, no memory will be allocated.
```
This doesn't mention that 0 is returned.

pshipton commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6134 @@
+	/**
+	 * Reallocates a block and adds it to a doubly-linked list for tracking purposes.  Actual size of the block is
+	 * requested size plus space for two pointers.
+	 * If size passed is 0, no memory will be allocated.
```
Doesn't mention that 0 is returned.

pshipton commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6134 @@
+	}
+
+	/**
+	 * Set object at specified index in parameter
```
Set the byte at the specified index and size

pshipton commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6134 @@
+	}
+
+	/**
+	 * Set object at specified index in obj parameter
```
Set the byte at the specified index and size in obj parameter

pshipton replied:
Sentences should end with a period. There are many single line sentences in this class which do not.

theresa-m replied:
Are periods required for sentences describing @param?

pshipton replied:
Looking at the OpenJDK javadoc it seems not
https://docs.oracle.com/javase/9/docs/api/java/lang/Appendable.html#append-java.lang.CharSequence-

pshipton commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6134 @@
+	}
+
+	/**
+	 * Swap specified bytes between source object and destination
```
What happens when srcSize and destSize are different?

theresa-m replied:
Currently nothing. The `copySwapMemory0` native is only a stub implementation so there isn't much to go on.

pshipton replied:
Hmm, I wonder if the javadoc should mention the method is currently unimplemented and will assert if called. @DanHeidinga thoughts?

pshipton commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6134 @@
+	}
+
+	/**
+	 * Swap specified bytes between source address and destination
```
What happens when srcSize and destSize are different?

pshipton commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6134 @@
+	}
+
+	/**
+	 * Returns byte offset to start of class or interface
```
Is the class or interface part correct? I thought its the offset to where the field is stored in the object.

pshipton commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6134 @@
+	}
+
+	/**
+	 * Returns static class or interface
```
Its hard to tell what this does from the description. Looking at the native code it returns the declaringClass of the static field.

pshipton commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6134 @@
+	}
+
+	/**
+	 * Return base offset of class array parameter
```
Its hard to tell what this function does from the description. It returns the offset in the array object at which the array data storage begins.

pshipton commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6134 @@
+
+	/**
+	 * Creates a class out of a given array of bytes with a specified ProtectionDomain.
+	 * A null ProtectionDomain will need to be resolved before the class can be used.
```
I'm not sure what this sentence means.

theresa-m replied:
What I meant by this was the spec for `ClassLoader.defineClass` says:
"If the given ProtectionDomain is null, then a default protection domain will be assigned to the class as specified in the documentation for defineClass(String, byte[], int, int). Before the class can be used it must be resolved."

Ours sets the protectionDomain to null if it is not defined so I'm not sure if this statement still applies.

pshipton replied:
Assigning a default ProtectionDomain, and resolving the class are two unrelated things.

pshipton commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6134 @@
+	 * 
+	 * @throws NullPointerException if b array of data is null
+	 * @throws ArrayIndexOutOfBoundsException if bLength is negative
+	 * @throws IndexOutOfBoundsException if offset or bLength is negative, or if 
```
bLength is negative is covered by the ArrayIndexOutOfBoundsException

theresa-m commented:
Ready for another review.

pshipton commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6132 @@
+	 * 
+	 * @param c class of same type as desired array
+	 * @param length desired length of array
+	 * @return allocated array of desired length and type, or null 
```
"or null is type os not a primitive" doesn't match the actual behavior

theresa-m replied:
"or null if class type is not a primitive wrapper" ?

pshipton replied:
Can it ever return null? It throws IllegalArgumentException.

theresa-m replied:
Oh right. No it cannot.

pshipton commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6132 @@
+	 * matches the existing value in the object.
+	 * The get operation has memory semantics of get.
+	 * The set operation has the memory semantics of set.
+	 *
```
This is the same as Plain, should it be different?

theresa-m replied:
They are the same in `VarHandle`.

pshipton commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6132 @@
+	 * matches the existing value in the object.
+	 * The get operation has memory semantics of get.
+	 * The set operation has the memory semantics of set.
+	 * 
```
This has the same doc as the weak variant, should they be different?

theresa-m replied:
These should have getVolatile and setVolatile for memory semantics instead of get and set.

pshipton commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6132 @@
+	 *
+	 * @param obj object into which to store the value
+	 * @param offset location to compare and store value in obj
+	 * @param compareValue value that is expected to be in obj
```
Should be getAcquire

pshipton replied:
Should be setRelease

pshipton replied:
Should be getByteAcquire. Please check the rest from here down.

pshipton commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6132 @@
+	 *
+	 * @param obj object into which to store the value
+	 * @param offset location to compare and store value in obj
+	 * @param compareValue value that is expected to be in obj
```
Should be weakCompareAndSetByteRelease

theresa-m replied:
This one was wrong but the others are intentional. `getAndSetXXXAcquire` and `getAndAddXXXAcquire` use `getAcquire` but `getAndBitwiseXXXXXXAcquire` just uses `get`.

pdbain-ibm commented:
@theresa-m would you kindly expedite this? It is required for https://github.com/eclipse/openj9/pull/579 which fixes a non-compliance.  Thanks.

theresa-m commented:
@pshipton ready for more review

theresa-m commented:
@andrewcraik can you also review this?

pdbain-ibm commented:
Reviewers: this is blocking Add Proxy class's package to the defining classloader #579.  Would you kindly review?

theresa-m commented:
Updated SPDX-License-Identifier for Unsafe.java

pshipton commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6131 @@
+	 * The set operation has the memory semantics of set.
+	 *
+	 * @param obj object into which to store the value
+	 * @param offset location to compare and store value in obj
```
What is this checking for? I don't understand the `>> 32`

pshipton replied:
This question is about the implementation of checkNativeAddress()

theresa-m replied:
This is checking that  the long type address parameter does not contain a number that exceeds 32 bits if the address size is 4 bytes. I'm updating the description to reflect that.

pshipton replied:
Isn't that what is32BitClean() is doing? I don't understand what checkNativeAddress() is doing.

theresa-m replied:
`checkNativeAddress()` passes with negative sign extended numbers while `is32BitClean()` would fail if passed a negative value because it uses logical shift right. I think this is to support sign extended pointers.

pshipton replied:
why the `& -2L`. This will allow many numbers greater than 32-bits to pass the test.

theresa-m replied:
`shiftedValue` after line 5708 will be either -1 if the address is 32 bits and negative or 0 if the address is 32 bits and positive.

`(shiftedValue + 1) & -2L` will only equate to zero if shiftedValue is either -1 or 0.

pshipton replied:
Ok, please add that as a comment.

theresa-m replied:
Okay. Done.

pshipton commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6131 @@
+	 * The set operation has the memory semantics of set.
+	 *
+	 * @param obj object into which to store the value
+	 * @param offset location to compare and store value in obj
```
Shouldn't it be `(BYTE_OFFSET_MASK & offset) != 0`?

pshipton replied:
nm

pshipton commented:
Created Eclipse CQ https://dev.eclipse.org/ipzilla/show_bug.cgi?id=15294

pshipton commented:
The CQ is approved.

pdbain-ibm commented:
When is this PR expected to be merged?  Thanks.

theresa-m commented:
@pdbain-ibm I believe we are waiting on @DanHeidinga for final review. @pshipton ?

pshipton commented:
> I believe we are waiting on @DanHeidinga for final review

correct

adamfarley commented:
Hi All,

I've taken the task of removing the extensions from Unsafe.java after this PR has been merged. These are the extensions that add the DBB native memory allocation methods to OpenJDK's Unsafe.java.

Everything is in place in the various extensions repositories, but I noticed that the Unsafe.java in JDK8 is in sun/misc, rather than jdk/internal/misc

```
Specifically:
JDK8 = jdk/src/share/classes/sun/misc/Unsafe.java
JDK9 = jdk/src/java.base/share/classes/jdk/internal/misc/Unsafe.java
JDK10+ = src/java.base/share/classes/jdk/internal/misc/Unsafe.java
```

I see no explicit reference to JDK8 in the PR, and I was wondering if there was a plan for that. If not, we'll be unable to undo the JDK8 Unsafe.java extensions until after we figure things out.

pshipton commented:
This change only affects Java 9 and later. I've added the jdk9 tag which was missing.

DanHeidinga commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6140 @@
+
+		ADDRESS_SIZE = theUnsafe.addressSize0();
+		IS_BIG_ENDIAN = theUnsafe.isBigEndian0();
+		UNALIGNED_ACCESS = theUnsafe.unalignedAccess0();
```
The native unconditionally returns TRUE.  Better to hardcode that here and avoid a call.  Can always be changed if we support a platform that doesn't allow unaligned access in the future.

DanHeidinga commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6140 @@
+	public native void fullFence();
+	
+	/* 
+	 * sun.misc.Unsafe natives 
```
`sun.misc` -> `jdk.internal.misc`

DanHeidinga commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6140 @@
+	 */
+	
+	/* 
+	 * Allocates a block and adds it to a doubly-linked list for tracking purposes.  Actual 
```
`doubly-linked list for tracking purposes` is that an OpenJDK or OpenJ9 statement?

theresa-m replied:
I took this from OpenJ9 `runtime/jcl/common/unsafe_mem.c:unsafeAllocateMemory`

DanHeidinga commented on jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java:
```diff
@@ -0,0 +1,6140 @@
+	 * The set operation has the memory semantics of set.
+	 *
+	 * @param obj object into which to store the value
+	 * @param offset location to compare and store value in obj
```
It's always better to have this kind of method throw the exception itself, and to mark the callers as throwing the return value, to cover cases were someone forgets the throw.

Having the caller also throw the return value makes it immediately obvious that the no further execution is possible.

DanHeidinga replied:
`Class` -> `Class<?>`

DanHeidinga replied:
bytes, ints and longs are usually more common than booleans, doubles, etc.  Can we reorder this to put the more common cases first?

pshipton commented:
> Is there a separate PR with tests?

#524

theresa-m commented:
@DanHeidinga  just pushed. Ready for another review.

theresa-m commented:
^ fixing copyright check

DanHeidinga commented:
Jenkins test all jdk8

DanHeidinga commented:
jenkins test sanity all jdk8

DanHeidinga commented:
launched the wrong build - should have launched the Java 9 versions.

jenkins test sanity all jdk9
jenkins test extended all jdk9

DanHeidinga commented:
jenkins test sanity all jdk9

DanHeidinga commented:
Some of the VarHandle tests are failing with this change:
```
00:02:51 [Test]         java.lang.AssertionError: expected [false] but found [true]
00:02:51 [Test]         	at org.testng.Assert.fail(Assert.java:94)
00:02:51 [Test]         	at org.testng.Assert.failNotEquals(Assert.java:513)
00:02:51 [Test]         	at org.testng.Assert.assertFalse(Assert.java:63)
00:02:51 [Test]         	at org.testng.Assert.assertFalse(Assert.java:73)
00:02:51 [Test]         	at org.openj9.test.varhandle.ArrayVarHandleTests.testBoolean(ArrayVarHandleTests.java:1425)
00:02:51 [Test]         	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(java.base@9-internal/Native Method)
00:02:51 [Test]         	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(java.base@9-internal/NativeMethodAccessorImpl.java:62)
00:02:51 [Test]         	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(java.base@9-internal/DelegatingMethodAccessorImpl.java:43)
00:02:51 [Test]         	at java.lang.reflect.Method.invoke(java.base@9-internal/Method.java:564)
00:02:51 [Test]         	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
00:02:51 [Test]         	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
00:02:51 [Test]         	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
00:02:51 [Test]         	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
00:02:51 [Test]         	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
00:02:51 [Test]         	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
00:02:51 [Test]         	at org.testng.TestRunner.privateRun(TestRunner.java:756)
00:02:51 [Test]         	at org.testng.TestRunner.run(TestRunner.java:610)
00:02:51 [Test]         	at org.testng.SuiteRunner.runTest(SuiteRunner.java:387)
00:02:51 [Test]         	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:382)
00:02:51 [Test]         	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:340)
00:02:51 [Test]         	at org.testng.SuiteRunner.run(SuiteRunner.java:289)
00:02:51 [Test]         	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
00:02:51 [Test]         	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
00:02:51 [Test]         	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1293)
00:02:51 [Test]         	at org.testng.TestNG.runSuitesLocally(TestNG.java:1218)
00:02:51 [Test]         	at org.testng.TestNG.runSuites(TestNG.java:1133)
00:02:51 [Test]         	at org.testng.TestNG.run(TestNG.java:1104)
00:02:51 [Test]         	at org.testng.TestNG.privateMain(TestNG.java:1434)
00:02:51 [Test]         	at org.testng.TestNG.main(TestNG.java:1403)
00:02:51 [Test]         
00:02:51 [Test]         FAILED: testByte
00:02:51 [Test]         java.lang.AssertionError: expected [3] but found [1]
00:02:51 [Test]         	at org.testng.Assert.fail(Assert.java:94)
00:02:51 [Test]         	at org.testng.Assert.failNotEquals(Assert.java:513)
00:02:51 [Test]         	at org.testng.Assert.assertEqualsImpl(Assert.java:135)
00:02:51 [Test]         	at org.testng.Assert.assertEquals(Assert.java:116)
00:02:51 [Test]         	at org.testng.Assert.assertEquals(Assert.java:326)
00:02:51 [Test]         	at org.testng.Assert.assertEquals(Assert.java:336)
00:02:51 [Test]         	at org.openj9.test.varhandle.ArrayVarHandleTests.testByte(ArrayVarHandleTests.java:106)
00:02:51 [Test]         	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(java.base@9-internal/Native Method)
00:02:51 [Test]         	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(java.base@9-internal/NativeMethodAccessorImpl.java:62)
00:02:51 [Test]         	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(java.base@9-internal/DelegatingMethodAccessorImpl.java:43)
00:02:51 [Test]         	at java.lang.reflect.Method.invoke(java.base@9-internal/Method.java:564)
00:02:51 [Test]         	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
00:02:51 [Test]         	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
00:02:51 [Test]         	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
00:02:51 [Test]         	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
00:02:51 [Test]         	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
00:02:51 [Test]         	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
00:02:51 [Test]         	at org.testng.TestRunner.privateRun(TestRunner.java:756)
00:02:51 [Test]         	at org.testng.TestRunner.run(TestRunner.java:610)
00:02:51 [Test]         	at org.testng.SuiteRunner.runTest(SuiteRunner.java:387)
00:02:51 [Test]         	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:382)
00:02:51 [Test]         	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:340)
00:02:51 [Test]         	at org.testng.SuiteRunner.run(SuiteRunner.java:289)
00:02:51 [Test]         	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
00:02:51 [Test]         	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
00:02:51 [Test]         	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1293)
00:02:51 [Test]         	at org.testng.TestNG.runSuitesLocally(TestNG.java:1218)
00:02:51 [Test]         	at org.testng.TestNG.runSuites(TestNG.java:1133)
00:02:51 [Test]         	at org.testng.TestNG.run(TestNG.java:1104)
00:02:51 [Test]         	at org.testng.TestNG.privateMain(TestNG.java:1434)
00:02:51 [Test]         	at org.testng.TestNG.main(TestNG.java:1403)
00:02:51 [Test]         
00:02:51 [Test]         FAILED: testChar
00:02:51 [Test]         java.lang.AssertionError: expected [3] but found [1]
00:02:51 [Test]         	at org.testng.Assert.fail(Assert.java:94)
00:02:51 [Test]         	at org.testng.Assert.failNotEquals(Assert.java:513)
00:02:51 [Test]         	at org.testng.Assert.assertEqualsImpl(Assert.java:135)
00:02:51 [Test]         	at org.testng.Assert.assertEquals(Assert.java:116)
00:02:51 [Test]         	at org.testng.Assert.assertEquals(Assert.java:347)
00:02:51 [Test]         	at org.testng.Assert.assertEquals(Assert.java:357)
00:02:51 [Test]         	at org.openj9.test.varhandle.ArrayVarHandleTests.testChar(ArrayVarHandleTests.java:256)
00:02:51 [Test]         	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(java.base@9-internal/Native Method)
00:02:51 [Test]         	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(java.base@9-internal/NativeMethodAccessorImpl.java:62)
00:02:51 [Test]         	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(java.base@9-internal/DelegatingMethodAccessorImpl.java:43)
00:02:51 [Test]         	at java.lang.reflect.Method.invoke(java.base@9-internal/Method.java:564)
00:02:51 [Test]         	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
00:02:51 [Test]         	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
00:02:51 [Test]         	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
00:02:51 [Test]         	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
00:02:51 [Test]         	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
00:02:51 [Test]         	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
00:02:51 [Test]         	at org.testng.TestRunner.privateRun(TestRunner.java:756)
00:02:51 [Test]         	at org.testng.TestRunner.run(TestRunner.java:610)
00:02:51 [Test]         	at org.testng.SuiteRunner.runTest(SuiteRunner.java:387)
00:02:51 [Test]         	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:382)
00:02:51 [Test]         	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:340)
00:02:51 [Test]         	at org.testng.SuiteRunner.run(SuiteRunner.java:289)
00:02:51 [Test]         	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
00:02:51 [Test]         	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
00:02:51 [Test]         	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1293)
00:02:51 [Test]         	at org.testng.TestNG.runSuitesLocally(TestNG.java:1218)
00:02:51 [Test]         	at org.testng.TestNG.runSuites(TestNG.java:1133)
00:02:51 [Test]         	at org.testng.TestNG.run(TestNG.java:1104)
00:02:51 [Test]         	at org.testng.TestNG.privateMain(TestNG.java:1434)
00:02:51 [Test]         	at org.testng.TestNG.main(TestNG.java:1403)
00:02:51 [Test]         
00:02:51 [Test]         FAILED: testShort
00:02:51 [Test]         java.lang.AssertionError: expected [3] but found [1]
00:02:51 [Test]         	at org.testng.Assert.fail(Assert.java:94)
00:02:51 [Test]         	at org.testng.Assert.failNotEquals(Assert.java:513)
00:02:51 [Test]         	at org.testng.Assert.assertEqualsImpl(Assert.java:135)
00:02:51 [Test]         	at org.testng.Assert.assertEquals(Assert.java:116)
00:02:51 [Test]         	at org.testng.Assert.assertEquals(Assert.java:368)
00:02:51 [Test]         	at org.testng.Assert.assertEquals(Assert.java:378)
00:02:51 [Test]         	at org.openj9.test.varhandle.ArrayVarHandleTests.testShort(ArrayVarHandleTests.java:1289)
00:02:51 [Test]         	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(java.base@9-internal/Native Method)
00:02:51 [Test]         	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(java.base@9-internal/NativeMethodAccessorImpl.java:62)
00:02:51 [Test]         	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(java.base@9-internal/DelegatingMethodAccessorImpl.java:43)
00:02:51 [Test]         	at java.lang.reflect.Method.invoke(java.base@9-internal/Method.java:564)
00:02:51 [Test]         	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
00:02:51 [Test]         	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
00:02:51 [Test]         	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
00:02:51 [Test]         	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
00:02:51 [Test]         	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
00:02:51 [Test]         	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
00:02:51 [Test]         	at org.testng.TestRunner.privateRun(TestRunner.java:756)
00:02:51 [Test]         	at org.testng.TestRunner.run(TestRunner.java:610)
00:02:51 [Test]         	at org.testng.SuiteRunner.runTest(SuiteRunner.java:387)
00:02:51 [Test]         	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:382)
00:02:51 [Test]         	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:340)
00:02:51 [Test]         	at org.testng.SuiteRunner.run(SuiteRunner.java:289)
00:02:51 [Test]         	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
00:02:51 [Test]         	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
00:02:51 [Test]         	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1293)
00:02:51 [Test]         	at org.testng.TestNG.runSuitesLocally(TestNG.java:1218)
00:02:51 [Test]         	at org.testng.TestNG.runSuites(TestNG.java:1133)
00:02:51 [Test]         	at org.testng.TestNG.run(TestNG.java:1104)
00:02:51 [Test]         	at org.testng.TestNG.privateMain(TestNG.java:1434)
00:02:51 [Test]         	at org.testng.TestNG.main(TestNG.java:1403)
00:02:51 [Test]         
00:02:51 [Test]         
00:02:51 [Test]         ===============================================
00:02:51 [Test]             VarHandle
00:02:51 [Test]             Tests run: 169, Failures: 4, Skips: 0
00:02:51 [Test]         ===============================================
00:02:51 [Test]         
```

theresa-m commented:
Working on a fix for that. Looks like I've misunderstood what should be returned on a failure.

theresa-m commented:
The varhandles tests should pass now.

DanHeidinga commented:
> Working on a fix for that. Looks like I've misunderstood what should be returned on a failure.

Can you point me at what you changed?  With a single commit I can't tell what's different now vs the last reviewed version.

DanHeidinga commented:
jenkins test extended all jdk9

DanHeidinga commented:
jenkins test sanity all jdk9

theresa-m commented:
All the changes I just pushed are tracked in this commit:
https://github.com/theresa-m/openj9/commit/84bd33f02c35a0f2edf764383be497caa592df29
This branch has the unsafe code and tests combined so its easier for me to write changes.

DanHeidinga commented:
LGTM

DanHeidinga commented:
@theresa-m Thanks for persisting through this change!  It's a major accomplishment to land this PR - implementing all of Unsafe with minimal test failures along the way is amazing.
